In [18]:
## 导入初始数据，Dow30的ret, rv, iv

import pandas as pd

# Load dow30_daily_returns_2021_2026.csv
dow30_df = pd.read_csv('dow30_daily_returns_2021_2026.csv')
print('First 5 rows of dow30_daily_returns_2021_2026.csv:')
display(dow30_df.head())

# Load merged_rv_data_filled.csv
rv_df = pd.read_csv('merged_rv_data_filled.csv')
print('\nFirst 5 rows of merged_rv_data_filled.csv:')
display(rv_df.head())

# Load merged_iv_data_filled.csv
iv_df = pd.read_csv('merged_iv_data_filled.csv')
print('\nFirst 5 rows of merged_iv_data_filled.csv:')
display(iv_df.head())


First 5 rows of dow30_daily_returns_2021_2026.csv:


,Date,AAPL,AMGN,AMZN,AXP,BA,CAT,CRM,CSCO,CVX,...,MMM,MRK,MSFT,NKE,PG,TRV,UNH,VZ,V,WMT
0,2021-01-27,-0.007684,-0.029157,-0.028126,-0.017736,-0.039741,-0.035148,-0.006541,0.009722,-0.012981,...,0.060813,-0.039626,0.002453,-0.030200,-0.035389,-0.029295,-0.029495,-0.025455,-0.034850,-0.024879
1,2021-01-28,-0.034985,-0.013184,0.001559,0.042773,0.016492,0.020539,0.007652,-0.007877,0.009297,...,-0.017305,0.011029,0.025891,0.028240,0.015423,0.003584,0.015496,0.004353,0.016669,-0.000626
2,2021-01-29,-0.037421,-0.025509,-0.009705,-0.022779,-0.015413,-0.008137,-0.004150,-0.016762,-0.042911,...,-0.042307,-0.010909,-0.029172,-0.008387,-0.016493,-0.026429,-0.013515,-0.011197,-0.025073,-0.022678
3,2021-02-01,0.016520,-0.004142,0.042630,0.006623,0.008497,0.010282,0.012857,0.018394,0.014202,...,-0.005465,0.003763,0.033152,0.015869,0.005928,0.014013,0.001259,-0.008585,0.026442,-0.008684
4,2021-02-02,0.006337,0.000249,0.011104,0.031359,0.026042,0.042118,0.025125,0.009471,0.007522,...,0.001832,0.005300,-0.000584,0.028590,-0.001396,0.011722,0.014132,0.004053,0.021426,0.010770



First 5 rows of merged_rv_data_filled.csv:


,Date,AAPL,AMGN,AMZN,AXP,BA,CAT,CRM,CSCO,CVX,...,MMM,MRK,MSFT,NKE,PG,TRV,UNH,VZ,V,WMT
0,2021-01-26,29.53,20.85,21.73,26.58,25.13,23.51,20.41,16.56,25.63,...,19.39,18.83,19.78,20.06,15.14,21.39,24.34,13.12,20.24,13.72
1,2021-01-27,29.49,21.90,21.90,27.81,26.17,24.38,20.53,17.52,26.59,...,22.59,20.61,21.68,20.31,17.38,22.39,24.84,14.85,20.68,14.82
2,2021-01-28,29.78,34.76,21.97,28.53,25.88,24.92,20.54,17.77,26.65,...,22.85,20.87,22.45,21.37,18.52,22.51,25.67,15.55,21.01,14.95
3,2021-01-29,31.28,35.11,21.91,28.73,27.13,26.28,21.16,18.02,26.84,...,24.74,21.34,22.91,21.66,18.76,22.96,26.08,15.53,23.27,15.70
4,2021-02-01,31.72,35.26,22.98,28.83,27.26,26.57,21.56,18.30,27.50,...,24.95,21.12,24.42,21.99,18.69,23.02,25.94,15.80,23.44,15.94



First 5 rows of merged_iv_data_filled.csv:


,Date,AAPL,AMGN,AMZN,AXP,BA,CAT,CRM,CSCO,CVX,...,MMM,MRK,MSFT,NKE,PG,TRV,UNH,VZ,V,WMT
0,2021-01-26,48.21,27.84,40.08,34.13,50.82,35.00,42.04,32.36,39.50,...,25.67,25.34,34.08,28.98,21.09,26.24,29.56,18.38,29.36,28.46
1,2021-01-27,49.12,33.28,46.18,39.85,54.21,42.33,46.51,35.00,43.58,...,31.38,28.98,35.27,36.33,26.70,33.50,33.56,21.04,33.67,33.76
2,2021-01-28,42.39,37.44,42.19,39.07,50.82,36.56,44.49,33.82,37.46,...,28.03,29.42,30.48,30.93,21.79,28.37,30.53,19.81,31.04,33.20
3,2021-01-29,44.55,35.46,43.02,38.45,52.25,35.95,45.27,36.22,40.38,...,30.35,29.09,33.59,32.76,23.65,33.76,31.94,21.41,33.12,33.62
4,2021-02-01,38.48,33.30,40.35,35.47,49.12,32.86,42.25,34.44,38.33,...,29.04,29.64,30.07,31.79,21.70,29.20,29.94,21.45,29.20,33.44


## 构建邻接矩阵

In [19]:
import numpy as np
import pandas as pd
from sklearn.covariance import GraphicalLassoCV

def GLASSO_Precision(subret):
    """
    使用Graphical Lasso方法计算精度矩阵并构建邻接矩阵

    Args:
        subret: 股票收益率数据框，每列代表一只股票的收益率时间序列

    Returns:
        adj_df: 标准化后的邻接矩阵，用于表示股票间的关系网络
    """

    # 获取股票数量和股票代码
    n = subret.shape[1]
    tickers = subret.columns

    # 使用交叉验证的Graphical Lasso拟合数据，自动选择最优的正则化参数alpha
    cov = GraphicalLassoCV().fit(subret)
    print('Alpha in GLASSO: %.3f' % cov.alpha_)

    # 从精度矩阵中提取非零元素，构建二值化的相关性矩阵
    corr = cov.precision_ != 0
    print('Sparsity of Adj: %.3f' % corr.mean())

    # 去除对角线元素（自相关），得到邻接矩阵
    corr_adj = corr - np.identity(n)

    # 计算度矩阵的逆平方根，用于标准化邻接矩阵
    # 加上小常数1e-8防止除零错误
    d_sqrt_inv = np.diag(np.sqrt(1/(corr_adj.sum(1)+1e-8)))

    # 应用对称标准化：D^(-1/2) * A * D^(-1/2)
    # 这样可以确保邻接矩阵的特征值在合理范围内
    adj_df = pd.DataFrame(np.dot(np.dot(d_sqrt_inv, corr_adj), d_sqrt_inv),
                         columns=tickers, index=tickers)

    return adj_df

print("GLASSO_Precision function defined.")



# Prepare the data for GLASSO_Precision by dropping the 'Date' column
subret_data = dow30_df.drop(columns=['Date'])

# Calculate the adjacency matrix
adjacency_matrix = GLASSO_Precision(subret_data)

# Display the first few rows and columns of the adjacency matrix
print('\nAdjacency Matrix (first 5 rows and columns):')
display(adjacency_matrix)


GLASSO_Precision function defined.
Alpha in GLASSO: 0.000
Sparsity of Adj: 0.367

Adjacency Matrix (first 5 rows and columns):


,AAPL,AMGN,AMZN,AXP,BA,CAT,CRM,CSCO,CVX,DIS,...,MMM,MRK,MSFT,NKE,PG,TRV,UNH,VZ,V,WMT
AAPL,0.000000,0.000000,0.077152,0.061314,0.074125,0.000000,0.071429,0.071429,0.000000,0.064820,...,0.000000,0.000000,0.109109,0.066815,0.000000,0.000000,0.000000,0.0,0.077152,0.133631
AMGN,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.109109,0.288675,0.000000,0.000000,0.000000,0.000000,0.235702,0.0,0.000000,0.000000
AMZN,0.077152,0.000000,0.000000,0.066227,0.080064,0.000000,0.077152,0.077152,0.000000,0.070014,...,0.000000,0.000000,0.117851,0.072169,0.000000,0.000000,0.000000,0.0,0.000000,0.144338
AXP,0.061314,0.000000,0.066227,0.000000,0.063628,0.066227,0.061314,0.061314,0.081111,0.055641,...,0.061314,0.000000,0.000000,0.057354,0.000000,0.081111,0.132453,0.0,0.066227,0.000000
BA,0.074125,0.000000,0.080064,0.063628,0.000000,0.080064,0.074125,0.000000,0.098058,0.067267,...,0.000000,0.000000,0.000000,0.069338,0.000000,0.000000,0.000000,0.0,0.000000,0.000000
CAT,0.000000,0.000000,0.000000,0.066227,0.080064,0.000000,0.000000,0.077152,0.102062,0.070014,...,0.077152,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.0,0.000000,0.000000
CRM,0.071429,0.000000,0.077152,0.061314,0.074125,0.000000,0.000000,0.071429,0.000000,0.064820,...,0.000000,0.000000,0.109109,0.066815,0.000000,0.000000,0.154303,0.0,0.077152,0.000000
CSCO,0.071429,0.000000,0.077152,0.061314,0.000000,0.077152,0.071429,0.000000,0.000000,0.064820,...,0.071429,0.000000,0.109109,0.066815,0.000000,0.000000,0.000000,0.0,0.000000,0.000000
CVX,0.000000,0.000000,0.000000,0.081111,0.098058,0.102062,0.000000,0.000000,0.000000,0.085749,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.125000,0.000000,0.0,0.000000,0.000000
DIS,0.064820,0.000000,0.070014,0.055641,0.067267,0.070014,0.064820,0.064820,0.085749,0.000000,...,0.064820,0.000000,0.000000,0.060634,0.000000,0.000000,0.000000,0.0,0.070014,0.000000


## Data Preprocessing for HAR Models

Prepare data for HAR-based volatility forecasting. The HAR model uses three components:
- **Daily**: 1-day lagged volatility
- **Weekly**: 5-day average lagged volatility  
- **Monthly**: 22-day average lagged volatility

In [20]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader, TensorDataset
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

# Set random seeds for reproducibility
np.random.seed(42)
torch.manual_seed(42)

print("Libraries imported successfully!")

# Prepare RV data (target variable)
rv_data = rv_df.set_index('Date')
print(f"RV data shape: {rv_data.shape}")
print(f"Date range: {rv_data.index[0]} to {rv_data.index[-1]}")

# Prepare return data (for adjacency matrix)
ret_data = dow30_df.set_index('Date')
print(f"Return data shape: {ret_data.shape}")
print(f"Date range: {ret_data.index[0]} to {ret_data.index[-1]}")

Libraries imported successfully!
RV data shape: (1256, 30)
Date range: 2021-01-26 to 2026-01-26
Return data shape: (1254, 30)
Date range: 2021-01-27 to 2026-01-23


In [ ]:
def create_HAR_features(rv_data, lags=[1, 5, 22]):
    """
    Create HAR (Heterogeneous Autoregressive) features
    
    Args:
        rv_data: DataFrame with RV values (Date index, stocks as columns)
        lags: List of lag periods [daily, weekly, monthly]
    
    Returns:
        DataFrame with HAR features for each stock
    """
    print(f"Input rv_data shape: {rv_data.shape}")
    print(f"Input rv_data columns: {rv_data.columns.tolist()}")
    print(f"Input rv_data index type: {type(rv_data.index)}")
    
    feature_list = []
    
    for ticker in rv_data.columns:
        print(f"Processing ticker: {ticker}")
        ticker_series = rv_data[ticker].copy()
        
        # Create DataFrame for this ticker
        ticker_df = pd.DataFrame()
        ticker_df['Date'] = rv_data.index
        ticker_df['Ticker'] = ticker
        ticker_df['Target'] = ticker_series.values
        
        # Create lagged features
        for lag in lags:
            # Calculate rolling mean of past 'lag' periods
            lag_feature = ticker_series.rolling(window=lag, min_periods=1).mean().shift(1)
            ticker_df[f'lag_{lag}'] = lag_feature.values
        
        # Remove rows with NaN values (first few rows due to lags)
        ticker_df_clean = ticker_df.dropna()
        print(f"  {ticker}: {len(ticker_df_clean)} valid samples after removing NaN")
        
        if len(ticker_df_clean) > 0:
            feature_list.append(ticker_df_clean)
    
    if len(feature_list) == 0:
        print("ERROR: No valid data after processing!")
        return pd.DataFrame()
    
    # Combine all tickers
    df = pd.concat(feature_list, ignore_index=True)
    print(f"Combined data shape: {df.shape}")
    
    return df

# Create HAR features
print("Creating HAR features...")
har_data = create_HAR_features(rv_data)
print(f"HAR data shape: {har_data.shape}")

if len(har_data) > 0:
    print("\nFirst few rows:")
    display(har_data.head())
    print(f"\nFeature columns: {[col for col in har_data.columns if 'lag' in col]}")
else:
    print("ERROR: HAR data is empty!")
    print("Checking original data...")
    print(f"rv_data shape: {rv_data.shape}")
    print(f"rv_data head:")
    display(rv_data.head())

Creating HAR features...


HAR data shape: (0, 6)

First few rows:


,Date,Ticker,Target,lag_1,lag_5,lag_22



Feature columns: ['lag_1', 'lag_5', 'lag_22']


## Train-Test Split

We'll use a time-based split:
- **Training**: First 80% of the data
- **Testing**: Last 20% of the data

In [ ]:
# Get unique dates
dates = sorted(har_data['Date'].unique())
print(f"Total dates: {len(dates)}")
print(f"Sample dates: {dates[:5] if len(dates) > 0 else 'No dates found'}")

# Check if we have data
if len(dates) == 0:
    print("ERROR: No dates found in har_data!")
    print(f"har_data shape: {har_data.shape}")
    print(f"har_data columns: {har_data.columns.tolist()}")
    display(har_data.head())
else:
    # 80-20 split
    split_idx = int(len(dates) * 0.8)
    if split_idx == 0:
        split_idx = max(1, len(dates) - 1)  # Ensure at least 1 training sample
    
    train_dates = dates[:split_idx]
    test_dates = dates[split_idx:]
    
    print(f"Training dates: {len(train_dates)} ({train_dates[0]} to {train_dates[-1]})")
    print(f"Testing dates: {len(test_dates)} ({test_dates[0]} to {test_dates[-1]})")
    
    # Split data
    train_data = har_data[har_data['Date'].isin(train_dates)].copy()
    test_data = har_data[har_data['Date'].isin(test_dates)].copy()
    
    print(f"\nTraining samples: {len(train_data)}")
    print(f"Testing samples: {len(test_data)}")
    
    # Prepare feature columns
    feature_cols = [col for col in har_data.columns if 'lag' in col]
    print(f"\nFeature columns: {feature_cols}")

Total dates: 0
Sample dates: No dates found
ERROR: No dates found in har_data!
har_data shape: (0, 6)
har_data columns: ['Date', 'Ticker', 'Target', 'lag_1', 'lag_5', 'lag_22']


,Date,Ticker,Target,lag_1,lag_5,lag_22


: 

## Model 1: HAR (Baseline Model)

The standard HAR model without graph structure. Each stock is modeled independently using its own lagged volatilities.

In [ ]:
# Train HAR model (Linear Regression)
print("Training HAR model...")

X_train = train_data[feature_cols].values
y_train = train_data['Target'].values

X_test = test_data[feature_cols].values
y_test = test_data['Target'].values

# Fit linear regression
har_model = LinearRegression()
har_model.fit(X_train, y_train)

# Make predictions
y_train_pred_har = har_model.predict(X_train)
y_test_pred_har = har_model.predict(X_test)

# Handle negative predictions (volatility must be positive)
y_train_pred_har = np.maximum(y_train_pred_har, 0.01)
y_test_pred_har = np.maximum(y_test_pred_har, 0.01)

print("HAR model training completed!")
print(f"Coefficients: {har_model.coef_}")
print(f"Intercept: {har_model.intercept_}")

Training HAR model...


NameError: name 'train_data' is not defined

## Model 2: GHAR (Graph HAR with GLASSO Adjacency)

GHAR extends HAR by incorporating cross-sectional dependencies through the GLASSO adjacency matrix. The model applies graph convolution to capture relationships between stocks.

In [ ]:
def apply_graph_convolution(df, adj_matrix, feature_cols):
    """
    Apply graph convolution: multiply adjacency matrix with features
    
    Args:
        df: DataFrame with features
        adj_matrix: Adjacency matrix (stocks x stocks)
        feature_cols: List of feature column names
    
    Returns:
        DataFrame with graph-convolved features
    """
    result_list = []
    
    for date in df['Date'].unique():
        date_df = df[df['Date'] == date].copy()
        
        # Get features for all stocks at this date (N x F matrix)
        features = date_df[feature_cols].values
        
        # Apply graph convolution: A @ X  (adjacency @ features)
        conv_features = adj_matrix.values @ features
        
        # Create new dataframe with convolved features
        conv_df = pd.DataFrame(
            conv_features,
            columns=[f'gc_{col}' for col in feature_cols]
        )
        conv_df['Date'] = date
        conv_df['Ticker'] = date_df['Ticker'].values
        conv_df['Target'] = date_df['Target'].values
        
        # Concatenate original and convolved features
        for col in feature_cols:
            conv_df[col] = date_df[col].values
        
        result_list.append(conv_df)
    
    return pd.concat(result_list, ignore_index=True)

# Apply graph convolution to training and test data
print("Applying graph convolution with GLASSO adjacency matrix...")
train_data_ghar = apply_graph_convolution(train_data, adjacency_matrix, feature_cols)
test_data_ghar = apply_graph_convolution(test_data, adjacency_matrix, feature_cols)

print(f"GHAR training data shape: {train_data_ghar.shape}")
print(f"GHAR testing data shape: {test_data_ghar.shape}")

# New feature columns include both original and graph-convolved features
ghar_feature_cols = [col for col in train_data_ghar.columns if 'lag' in col or 'gc_' in col]
print(f"\nGHAR feature columns ({len(ghar_feature_cols)}): {ghar_feature_cols}")

In [ ]:
# Train GHAR model
print("Training GHAR model...")

X_train_ghar = train_data_ghar[ghar_feature_cols].values
y_train_ghar = train_data_ghar['Target'].values

X_test_ghar = test_data_ghar[ghar_feature_cols].values
y_test_ghar = test_data_ghar['Target'].values

# Fit linear regression with graph features
ghar_model = LinearRegression()
ghar_model.fit(X_train_ghar, y_train_ghar)

# Make predictions
y_train_pred_ghar = ghar_model.predict(X_train_ghar)
y_test_pred_ghar = ghar_model.predict(X_test_ghar)

# Handle negative predictions
y_train_pred_ghar = np.maximum(y_train_pred_ghar, 0.01)
y_test_pred_ghar = np.maximum(y_test_pred_ghar, 0.01)

print("GHAR model training completed!")
print(f"Number of coefficients: {len(ghar_model.coef_)}")
print(f"Intercept: {ghar_model.intercept_:.4f}")

## Model 3: GNNHAR (Graph Neural Network HAR)

GNNHAR uses neural networks with graph convolutional layers to model non-linear relationships in volatility forecasting.

In [ ]:
# Define Graph Convolutional Layer
class GraphConvLayer(nn.Module):
    """Graph Convolutional Layer: H' = A @ H @ W + b"""
    def __init__(self, in_features, out_features, bias=True):
        super(GraphConvLayer, self).__init__()
        self.weight = nn.Parameter(torch.FloatTensor(in_features, out_features))
        nn.init.xavier_uniform_(self.weight, gain=nn.init.calculate_gain('relu'))
        
        if bias:
            self.bias = nn.Parameter(torch.FloatTensor(1, out_features))
            nn.init.ones_(self.bias)
        else:
            self.bias = None
    
    def forward(self, node_feature, adj):
        """
        Args:
            node_feature: [batch_size, N, in_features]
            adj: [N, N] adjacency matrix
        Returns:
            [batch_size, N, out_features]
        """
        h = torch.matmul(node_feature, self.weight)  # H @ W
        output = torch.matmul(adj, h)  # A @ (H @ W)
        if self.bias is not None:
            return output + self.bias
        return output


# Define GNNHAR1L Model (1-layer GNN)
class GNNHAR1L(nn.Module):
    """
    GNNHAR with 1 graph convolutional layer
    Architecture: Linear HAR branch + GCN branch with residual connection
    """
    def __init__(self, n_hid=16):
        super(GNNHAR1L, self).__init__()
        
        # Linear HAR component
        self.linear1 = nn.Linear(3, 1, bias=True)
        
        # Graph neural network component
        self.gcn1 = GraphConvLayer(3, n_hid, bias=False)
        self.mlp1 = nn.Linear(n_hid, 1, bias=False)
        self.relu = nn.ReLU()
    
    def forward(self, node_feat, adj):
        """
        Args:
            node_feat: [batch_size, N, 3] - HAR features (daily, weekly, monthly)
            adj: [N, N] - adjacency matrix
        Returns:
            [batch_size, N] - predicted volatilities
        """
        # Linear HAR branch
        H1 = self.linear1(node_feat)  # [batch_size, N, 1]
        
        # GNN branch
        H2 = self.gcn1(node_feat, adj)  # [batch_size, N, n_hid]
        H2 = self.relu(H2)
        H2 = self.mlp1(H2)  # [batch_size, N, 1]
        
        # Residual connection
        res = H1 + H2
        res = self.relu(res)  # Ensure positive volatility
        
        return res.squeeze(-1)  # [batch_size, N]

print("GNNHAR model classes defined!")

In [ ]:
# Prepare data for GNNHAR (batch format)
def prepare_gnnhar_data(df, feature_cols, n_stocks):
    """
    Prepare data in batch format for GNNHAR
    Each sample is all stocks at a single date
    
    Args:
        df: DataFrame with features
        feature_cols: List of feature columns
        n_stocks: Number of stocks
    
    Returns:
        X: [n_dates, n_stocks, n_features]
        y: [n_dates, n_stocks]
    """
    dates = sorted(df['Date'].unique())
    X_list = []
    y_list = []
    
    for date in dates:
        date_df = df[df['Date'] == date].sort_values('Ticker')
        if len(date_df) == n_stocks:  # Ensure we have all stocks
            X_list.append(date_df[feature_cols].values)
            y_list.append(date_df['Target'].values)
    
    X = np.array(X_list)  # [n_dates, n_stocks, n_features]
    y = np.array(y_list)  # [n_dates, n_stocks]
    
    return X, y

# Get number of stocks
n_stocks = len(rv_data.columns)
print(f"Number of stocks: {n_stocks}")

# Prepare data
X_train_gnn, y_train_gnn = prepare_gnnhar_data(train_data, feature_cols, n_stocks)
X_test_gnn, y_test_gnn = prepare_gnnhar_data(test_data, feature_cols, n_stocks)

print(f"X_train_gnn shape: {X_train_gnn.shape}")  # [n_dates, n_stocks, 3]
print(f"y_train_gnn shape: {y_train_gnn.shape}")  # [n_dates, n_stocks]
print(f"X_test_gnn shape: {X_test_gnn.shape}")
print(f"y_test_gnn shape: {y_test_gnn.shape}")

# Convert to PyTorch tensors
X_train_tensor = torch.FloatTensor(X_train_gnn)
y_train_tensor = torch.FloatTensor(y_train_gnn)
X_test_tensor = torch.FloatTensor(X_test_gnn)
y_test_tensor = torch.FloatTensor(y_test_gnn)

# Convert adjacency matrix to tensor
adj_tensor = torch.FloatTensor(adjacency_matrix.values)

print("\nData conversion completed!")

In [ ]:
# Initialize model
n_hid = 16
gnnhar_model = GNNHAR1L(n_hid=n_hid)
print(f"Model initialized with {n_hid} hidden units")
print(gnnhar_model)

# Training parameters
n_epochs = 500
learning_rate = 0.001
batch_size = 32

# Loss function and optimizer
criterion = nn.MSELoss()
optimizer = optim.Adam(gnnhar_model.parameters(), lr=learning_rate)

# Create DataLoader
train_dataset = TensorDataset(X_train_tensor, y_train_tensor)
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)

print(f"\nTraining configuration:")
print(f"  Epochs: {n_epochs}")
print(f"  Learning rate: {learning_rate}")
print(f"  Batch size: {batch_size}")
print(f"  Number of batches: {len(train_loader)}")

In [ ]:
# Training loop
print("Training GNNHAR model...")
train_losses = []
test_losses = []

for epoch in range(n_epochs):
    gnnhar_model.train()
    epoch_loss = 0
    
    for batch_X, batch_y in train_loader:
        # Forward pass
        predictions = gnnhar_model(batch_X, adj_tensor)
        loss = criterion(predictions, batch_y)
        
        # Backward pass
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        
        epoch_loss += loss.item()
    
    # Calculate average training loss
    avg_train_loss = epoch_loss / len(train_loader)
    train_losses.append(avg_train_loss)
    
    # Evaluate on test set
    gnnhar_model.eval()
    with torch.no_grad():
        test_pred = gnnhar_model(X_test_tensor, adj_tensor)
        test_loss = criterion(test_pred, y_test_tensor).item()
        test_losses.append(test_loss)
    
    # Print progress every 50 epochs
    if (epoch + 1) % 50 == 0:
        print(f"Epoch [{epoch+1}/{n_epochs}] - Train Loss: {avg_train_loss:.4f}, Test Loss: {test_loss:.4f}")

print("\nTraining completed!")

In [ ]:
# Plot training curves
plt.figure(figsize=(10, 5))
plt.plot(train_losses, label='Training Loss', alpha=0.7)
plt.plot(test_losses, label='Test Loss', alpha=0.7)
plt.xlabel('Epoch')
plt.ylabel('MSE Loss')
plt.title('GNNHAR Training Progress')
plt.legend()
plt.grid(True, alpha=0.3)
plt.tight_layout()
plt.show()

print(f"Final training loss: {train_losses[-1]:.4f}")
print(f"Final test loss: {test_losses[-1]:.4f}")

In [ ]:
# Generate predictions for GNNHAR
gnnhar_model.eval()
with torch.no_grad():
    y_train_pred_gnn = gnnhar_model(X_train_tensor, adj_tensor).numpy()
    y_test_pred_gnn = gnnhar_model(X_test_tensor, adj_tensor).numpy()

# Flatten predictions to match linear models format
y_train_pred_gnn_flat = y_train_pred_gnn.flatten()
y_test_pred_gnn_flat = y_test_pred_gnn.flatten()
y_train_gnn_flat = y_train_gnn.flatten()
y_test_gnn_flat = y_test_gnn.flatten()

print(f"GNNHAR predictions generated!")
print(f"Train predictions shape: {y_train_pred_gnn_flat.shape}")
print(f"Test predictions shape: {y_test_pred_gnn_flat.shape}")

## Error Analysis: MSE and QLIKE

We evaluate all models using two metrics:

1. **MSE (Mean Squared Error)**: $\text{MSE} = \frac{1}{n}\sum_{i=1}^{n}(y_i - \hat{y}_i)^2$

2. **QLIKE (Quasi-Likelihood)**: $\text{QLIKE} = \frac{1}{n}\sum_{i=1}^{n}\left(\frac{y_i}{\hat{y}_i} - \log\frac{y_i}{\hat{y}_i} - 1\right)$

QLIKE is specifically designed for volatility forecasting and is more robust to outliers than MSE.

In [ ]:
def calculate_mse(y_true, y_pred):
    """Calculate Mean Squared Error"""
    return np.mean((y_true - y_pred) ** 2)

def calculate_qlike(y_true, y_pred, epsilon=1e-4):
    """
    Calculate Quasi-Likelihood (QLIKE) loss for volatility forecasting
    QLIKE = mean(y_true/y_pred - log(y_true/y_pred) - 1)
    
    Args:
        y_true: True volatility values
        y_pred: Predicted volatility values
        epsilon: Small constant to avoid division by zero
    """
    # Ensure positive values and avoid division by zero
    y_pred = np.maximum(y_pred, epsilon)
    y_true = np.maximum(y_true, epsilon)
    
    # QLIKE formula
    ratio = y_true / y_pred
    qlike = np.mean(ratio - np.log(ratio) - 1)
    
    return qlike

print("Error metric functions defined!")

In [ ]:
# Calculate errors for all models
results = {
    'Model': [],
    'Train_MSE': [],
    'Test_MSE': [],
    'Train_QLIKE': [],
    'Test_QLIKE': []
}

# HAR Model
results['Model'].append('HAR')
results['Train_MSE'].append(calculate_mse(y_train, y_train_pred_har))
results['Test_MSE'].append(calculate_mse(y_test, y_test_pred_har))
results['Train_QLIKE'].append(calculate_qlike(y_train, y_train_pred_har))
results['Test_QLIKE'].append(calculate_qlike(y_test, y_test_pred_har))

# GHAR Model
results['Model'].append('GHAR')
results['Train_MSE'].append(calculate_mse(y_train_ghar, y_train_pred_ghar))
results['Test_MSE'].append(calculate_mse(y_test_ghar, y_test_pred_ghar))
results['Train_QLIKE'].append(calculate_qlike(y_train_ghar, y_train_pred_ghar))
results['Test_QLIKE'].append(calculate_qlike(y_test_ghar, y_test_pred_ghar))

# GNNHAR Model
results['Model'].append('GNNHAR1L')
results['Train_MSE'].append(calculate_mse(y_train_gnn_flat, y_train_pred_gnn_flat))
results['Test_MSE'].append(calculate_mse(y_test_gnn_flat, y_test_pred_gnn_flat))
results['Train_QLIKE'].append(calculate_qlike(y_train_gnn_flat, y_train_pred_gnn_flat))
results['Test_QLIKE'].append(calculate_qlike(y_test_gnn_flat, y_test_pred_gnn_flat))

# Create results DataFrame
results_df = pd.DataFrame(results)
print("\n" + "="*80)
print("MODEL COMPARISON - ERROR ANALYSIS")
print("="*80)
display(results_df)

# Calculate improvement over HAR baseline
print("\n" + "="*80)
print("IMPROVEMENT OVER HAR BASELINE (%)")
print("="*80)
har_test_mse = results_df[results_df['Model'] == 'HAR']['Test_MSE'].values[0]
har_test_qlike = results_df[results_df['Model'] == 'HAR']['Test_QLIKE'].values[0]

for idx, row in results_df.iterrows():
    if row['Model'] != 'HAR':
        mse_improve = (har_test_mse - row['Test_MSE']) / har_test_mse * 100
        qlike_improve = (har_test_qlike - row['Test_QLIKE']) / har_test_qlike * 100
        print(f"{row['Model']:12s}: MSE {mse_improve:+.2f}%  |  QLIKE {qlike_improve:+.2f}%")

In [ ]:
# Visualization of model comparison
fig, axes = plt.subplots(2, 2, figsize=(14, 10))

# Plot 1: Training MSE
ax1 = axes[0, 0]
ax1.bar(results_df['Model'], results_df['Train_MSE'], color=['#1f77b4', '#ff7f0e', '#2ca02c'])
ax1.set_ylabel('MSE')
ax1.set_title('Training MSE Comparison')
ax1.grid(True, alpha=0.3)

# Plot 2: Test MSE
ax2 = axes[0, 1]
ax2.bar(results_df['Model'], results_df['Test_MSE'], color=['#1f77b4', '#ff7f0e', '#2ca02c'])
ax2.set_ylabel('MSE')
ax2.set_title('Test MSE Comparison')
ax2.grid(True, alpha=0.3)

# Plot 3: Training QLIKE
ax3 = axes[1, 0]
ax3.bar(results_df['Model'], results_df['Train_QLIKE'], color=['#1f77b4', '#ff7f0e', '#2ca02c'])
ax3.set_ylabel('QLIKE')
ax3.set_title('Training QLIKE Comparison')
ax3.grid(True, alpha=0.3)

# Plot 4: Test QLIKE
ax4 = axes[1, 1]
ax4.bar(results_df['Model'], results_df['Test_QLIKE'], color=['#1f77b4', '#ff7f0e', '#2ca02c'])
ax4.set_ylabel('QLIKE')
ax4.set_title('Test QLIKE Comparison')
ax4.grid(True, alpha=0.3)

plt.tight_layout()
plt.show()

## Prediction Visualization

Visualize predictions vs actual values for selected stocks on the test set.

In [ ]:
# Select 4 stocks for visualization
selected_tickers = ['AAPL', 'MSFT', 'JPM', 'BA']
stocks = rv_data.columns.tolist()

fig, axes = plt.subplots(2, 2, figsize=(16, 10))
axes = axes.flatten()

for idx, ticker in enumerate(selected_tickers):
    if ticker in stocks:
        stock_idx = stocks.index(ticker)
        
        # Get test data for this stock
        test_dates_idx = range(len(test_dates))
        y_true_stock = y_test_gnn[:, stock_idx]
        y_pred_har_stock = y_test_pred_gnn[:, stock_idx]  # From GNNHAR for consistency
        
        # For HAR and GHAR, we need to extract stock-specific predictions
        test_data_stock = test_data[test_data['Ticker'] == ticker].copy()
        test_data_ghar_stock = test_data_ghar[test_data_ghar['Ticker'] == ticker].copy()
        
        ax = axes[idx]
        ax.plot(test_dates_idx, y_true_stock, label='Actual RV', 
                linewidth=2, alpha=0.8, color='black')
        
        # Plot predictions from GNNHAR
        ax.plot(test_dates_idx, y_pred_har_stock, label='GNNHAR1L', 
                linewidth=1.5, alpha=0.7, linestyle='--')
        
        ax.set_xlabel('Time (Test Period)')
        ax.set_ylabel('Realized Volatility')
        ax.set_title(f'{ticker} - Actual vs Predicted RV')
        ax.legend()
        ax.grid(True, alpha=0.3)

plt.tight_layout()
plt.show()

## Stock-Level Error Analysis

Compare model performance across individual stocks to identify which stocks are easier/harder to predict.

In [ ]:
# Calculate MSE for each stock
stock_mse_results = []

for i, ticker in enumerate(stocks):
    y_true_stock = y_test_gnn[:, i]
    
    # GNNHAR predictions
    y_pred_gnn_stock = y_test_pred_gnn[:, i]
    mse_gnn = calculate_mse(y_true_stock, y_pred_gnn_stock)
    qlike_gnn = calculate_qlike(y_true_stock, y_pred_gnn_stock)
    
    stock_mse_results.append({
        'Ticker': ticker,
        'Test_MSE': mse_gnn,
        'Test_QLIKE': qlike_gnn,
        'Mean_RV': np.mean(y_true_stock)
    })

stock_results_df = pd.DataFrame(stock_mse_results)
stock_results_df = stock_results_df.sort_values('Test_MSE')

print("\n" + "="*80)
print("STOCK-LEVEL PERFORMANCE (GNNHAR1L on Test Set)")
print("="*80)
print("\nTop 5 Best Predicted Stocks (Lowest MSE):")
display(stock_results_df.head())

print("\nTop 5 Worst Predicted Stocks (Highest MSE):")
display(stock_results_df.tail())

## Summary and Conclusions

### Model Comparison Summary

This notebook implements and compares three volatility forecasting models:

1. **HAR (Baseline)**: Standard Heterogeneous Autoregressive model
   - Uses only temporal dependencies (daily, weekly, monthly lags)
   - Linear model trained independently for each stock

2. **GHAR (Graph HAR)**: HAR with graph structure
   - Incorporates cross-sectional dependencies via GLASSO adjacency matrix
   - Applies graph convolution to capture inter-stock relationships
   - Still a linear model but with enhanced features

3. **GNNHAR1L (Graph Neural Network HAR)**: Deep learning approach
   - 1-layer graph convolutional neural network
   - Captures non-linear relationships in volatility dynamics
   - Residual connection combines linear HAR and GCN components

### Key Findings

The adjacency matrix constructed using Graphical Lasso (GLASSO):
- **Alpha = 0.000**: Minimal regularization needed
- **Sparsity = 0.367**: ~37% of possible edges exist
- This indicates moderate connectivity in the Dow 30 stock network

### Evaluation Metrics

- **MSE (Mean Squared Error)**: Standard metric for regression
- **QLIKE (Quasi-Likelihood)**: Specifically designed for volatility forecasting, more robust to outliers

### Advantages of Graph-Based Models

1. **Cross-sectional information**: Capture relationships between stocks
2. **Network effects**: Model how volatility spillovers across the market
3. **Improved predictions**: Especially for stocks with strong inter-dependencies

### Future Extensions

- Test with different adjacency matrix methods (correlation, sector-based)
- Experiment with deeper GNN architectures (GNNHAR2L, GNNHAR3L)
- Implement rolling window forecasting for more robust evaluation
- Add implied volatility (IV) as additional features

## 构建HAR